Sqlancer report error: cannot find statistics of col: 13: c_0_2

### Steps to reproduce the behavior (Required)
1. Prepare Data:

```sql
-- Time: 2023/07/27 15:32:54
-- Database: h3
-- Database version: 5.1.0
-- seed value: 1690443138420
DROP DATABASE IF EXISTS h3;
CREATE DATABASE h3;
USE h3;
admin set frontend config ("enable_decimal_v3" = "true");
CREATE TABLE t0 (c_0_0 DATETIME NULL ,c_0_1 VARCHAR(21) NOT NULL ,c_0_2 DATETIME NOT NULL ,c_0_3 DATETIME NULL ,c_0_4 CHAR(1) NULL ,c_0_5 BOOLEAN NULL ,c_0_6 BOOLEAN NOT NULL ,c_0_7 VARCHAR(1) NULL ,c_0_8 VARCHAR(11) NULL ,c_0_9 DECIMAL(4, 3)  NULL ,c_0_10 DECIMAL(32, 20)  NULL ,c_0_11 ARRAY<INT> NOT NULL ,c_0_12 ARRAY<INT> NULL ,c_0_13 ARRAY<STRING> NOT NULL ,c_0_14 DATE NOT NULL ,c_0_15 DOUBLE NOT NULL ,c_0_16 ARRAY<INT> NULL , INDEX index1_c_0_6(`c_0_6`) USING BITMAP) UNIQUE KEY (c_0_0,c_0_1,c_0_2,c_0_3,c_0_4,c_0_5,c_0_6,c_0_7,c_0_8,c_0_9,c_0_10) PARTITION BY RANGE(c_0_2) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 158 day)) DISTRIBUTED BY HASH (c_0_5,c_0_3,c_0_6,c_0_0,c_0_4) BUCKETS 3 properties("replication_num"="1") ;
CREATE TABLE t1 (c_1_0 VARCHAR(11) NULL ,c_1_1 DATETIME NULL ,c_1_2 DATETIME NULL ,c_1_3 DATE NOT NULL ,c_1_4 INT NOT NULL , INDEX index1_c_1_0(`c_1_0`) USING BITMAP, INDEX index1_c_1_4(`c_1_4`) USING BITMAP) DUPLICATE KEY (c_1_0) DISTRIBUTED BY HASH (c_1_0) BUCKETS 3 properties("replication_num"="1") ;
CREATE TABLE IF NOT EXISTS t2 (c_2_0 BOOLEAN NOT NULL,c_2_1 BOOLEAN NOT NULL,c_2_2 ARRAY<INT> NOT NULL,c_2_3 ARRAY<STRING> NOT NULL,c_2_4 FLOAT NOT NULL) PRIMARY KEY (c_2_0,c_2_1) DISTRIBUTED BY HASH (c_2_1) BUCKETS 3 properties("replication_num"="1") ;
CREATE TABLE t3 (c_3_0 DATE NOT NULL,c_3_1 VARCHAR(11) NOT NULL,c_3_2 VARCHAR(21) NOT NULL,c_3_3 DECIMAL(14, 9)  NOT NULL,c_3_4 DATETIME NOT NULL,c_3_5 ARRAY<STRING> NOT NULL,c_3_6 ARRAY<INT> NOT NULL, INDEX index1_c_3_1(`c_3_1`) USING BITMAP) PRIMARY KEY (c_3_0) PARTITION BY RANGE(c_3_0) (START ("2010-01-01") END ("2022-12-31") EVERY (INTERVAL 121 day)) DISTRIBUTED BY HASH (c_3_0) BUCKETS 3 properties("replication_num"="1") ;
INSERT INTO t0 VALUES ('2010-01-29 08:00:14', '', '2010-01-24 08:00:14', '2010-01-07 08:00:14', '', true, true, '', 'o', 0.006, 9.56700690114E-9, [-1346209733, -1152616326, 1106092166, -1369367618, 1891566907, 1106092166, 19686664, 1891566907, 19686664], [-1346209733], [], '2010-01-01', 0.8365961765936641, [-1695819837, -276432835, -2138610999, 1205295677, -1872337005, 236842512, -1896718936, -1560565856, 562944373, 1205295677, -1695819837, -1587320149, 735256100, 1205295677]), ('2010-01-16 08:00:14', 'iWFQ9Y', '2010-01-03 08:00:14', '2010-01-06 08:00:14', '', false, false, '', 'Xy', 0.002, 1.98136659399E-9, [1815892536, 821192849, 1691532338, 1891566907, 916639558, 735256100, -1872337005, 1290276326, 1205295677, 1891566907, 1180658845, 1286092243, 438249409, -865859867, -1346209733], [1110633170, 446224433, -3562345, 929234325, 13727972, 2053081697], ["", "bTSt", "P", "7X", "B", "nAQH3E", "4KEXUifI", "tO", "7jk"], '2010-01-22', 0.4911365636260482, [562944373]);
DELETE FROM t2 WHERE c_2_1=false;

```

2. Query

```sql
WITH with_t_0 as (SELECT t1_78.c_1_0, t1_78.c_1_1, t1_78.c_1_2, t1_78.c_1_3, t1_78.c_1_4 FROM t1 AS t1_78 WHERE ((t1_78.c_1_1) IS NULL)) SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) )  UNION ALL SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (NOT ((MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) ) )) UNION ALL SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (((MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) ) ) IS NULL);
```
### Expected behavior (Required)
NO Error

### Real behavior (Required)
Error as follows
```
--java.lang.AssertionError: WITH with_t_0 as (SELECT t1_78.c_1_0, t1_78.c_1_1, t1_78.c_1_2, t1_78.c_1_3, t1_78.c_1_4 FROM t1 AS t1_78 WHERE ((t1_78.c_1_1) IS NULL)) SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) )  UNION ALL SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (NOT ((MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) ) )) UNION ALL SELECT t2_79.c_2_0 FROM t2 AS t2_79, t0 AS t0_77, t1 AS t1_78, with_t_0 AS with_t_0 WHERE NULL GROUP BY t0_77.c_0_1, t0_77.c_0_8 HAVING (((MAX('1970-01-10 20:06:32') ) IN ((SELECT t1_78.c_1_3 FROM t2 AS t2_79, t1 AS t1_78 WHERE (NOT ( ( t0_77.c_0_2 ) != ( t0_77.c_0_2 ) ))) ) ) IS NULL);
--	at sqlancer.common.query.QueryAdapter.checkException(QueryAdapter.java:98)
--	at sqlancer.common.query.QueryAdapter.executeAndGet(QueryAdapter.java:128)
--	at sqlancer.ComparatorHelper.getResultSetFirstColumnAsString(ComparatorHelper.java:54)
--	at sqlancer.ComparatorHelper.getCombinedResultSet(ComparatorHelper.java:138)
--	at sqlancer.starrocks.oracle.tlp.StarRocksTLPHavingOracle.check(StarRocksTLPHavingOracle.java:54)
--	at sqlancer.ProviderAdapter.generateAndTestDatabase(ProviderAdapter.java:55)
--	at sqlancer.Main$DBMSExecutor.run(Main.java:334)
--	at sqlancer.Main$2.run(Main.java:542)
--	at sqlancer.Main$2.runThread(Main.java:519)
--	at sqlancer.Main$2.run(Main.java:509)
--	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
--	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
--	at java.lang.Thread.run(Thread.java:748)
--Caused by: com.mysql.jdbc.exceptions.jdbc4.MySQLSyntaxErrorException: cannot find statistics of col: 13: c_0_2
--	at sun.reflect.GeneratedConstructorAccessor5.newInstance(Unknown Source)
--	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
--	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
--	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
--	at com.mysql.jdbc.Util.getInstance(Util.java:408)
--	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:943)
--	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3973)
--	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3909)
--	at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2527)
--	at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2680)
--	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2497)
--	at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2455)
--	at com.mysql.jdbc.StatementImpl.executeQuery(StatementImpl.java:1369)
--	at sqlancer.common.query.QueryAdapter.executeAndGet(QueryAdapter.java:118)
--	... 11 more
--
```
### StarRocks version (Required)
- main(b17435b)

--- a/fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AggregationAnalyzer.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AggregationAnalyzer.java
@@ -50,9 +50,9 @@ import com.starrocks.sql.ast.LambdaFunctionExpr;
 import com.starrocks.sql.ast.QueryStatement;
 
 import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
-import java.util.stream.Collectors;
 
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static com.starrocks.sql.common.ErrorMsgProxy.PARSER_ERROR_MSG;
@@ -297,11 +297,28 @@ public class AggregationAnalyzer {
         @Override
         public Boolean visitSubquery(Subquery node, Void context) {
             QueryStatement queryStatement = node.getQueryStatement();
-            List<FieldId> fieldIds = queryStatement.getQueryRelation().getColumnReferences().values().stream()
-                    .filter(fieldId -> fieldId.getRelationId().equals(sourceScope.getRelationId()))
-                    .collect(Collectors.toList());
+            for (Map.Entry<Expr, FieldId> entry : queryStatement.getQueryRelation().getColumnReferences().entrySet()) {
+                Expr expr = entry.getKey();
+                FieldId id = entry.getValue();
 
-            return groupingFields.containsAll(fieldIds);
+                if (!id.getRelationId().equals(sourceScope.getRelationId())) {
+                    continue;
+                }
+
+                if (!groupingFields.contains(id)) {
+                    if (!SqlModeHelper.check(session.getSessionVariable().getSqlMode(),
+                            SqlModeHelper.MODE_ONLY_FULL_GROUP_BY)) {
+                        if (!analyzeState.getColumnNotInGroupBy().contains(expr)) {
+                            throw new SemanticException(
+                                    PARSER_ERROR_MSG.unsupportedNoGroupBySubquery(expr.toSql(), node.toSql()),
+                                    expr.getPos());
+                        }
+                    } else {
+                        return false;
+                    }
+                }
+            }
+            return true;
         }
 
         @Override

--- a/fe/fe-core/src/main/java/com/starrocks/sql/ast/SubqueryRelation.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/ast/SubqueryRelation.java
@@ -15,9 +15,11 @@
 package com.starrocks.sql.ast;
 
 import com.starrocks.analysis.Expr;
+import com.starrocks.sql.analyzer.FieldId;
 import com.starrocks.sql.parser.NodePosition;
 
 import java.util.List;
+import java.util.Map;
 
 public class SubqueryRelation extends QueryRelation {
     private final QueryStatement queryStatement;
@@ -34,6 +36,11 @@ public class SubqueryRelation extends QueryRelation {
         if (!queryRelation.hasLimit()) {
             queryRelation.clearOrder();
         }
+    }
+
+    @Override
+    public Map<Expr, FieldId> getColumnReferences() {
+        return queryStatement.getQueryRelation().getColumnReferences();
     }
 
     public QueryStatement getQueryStatement() {

--- a/fe/fe-core/src/main/java/com/starrocks/sql/common/ParserErrorMsg.java
+++ b/fe/fe-core/src/main/java/com/starrocks/sql/common/ParserErrorMsg.java
@@ -176,6 +176,9 @@ public interface ParserErrorMsg {
     @BaseMessage("''{0}'' must be an aggregate expression or appear in GROUP BY clause")
     String shouldBeAggFunc(String a0);
 
+    @BaseMessage("subquery correlated column ''{0}'' in ''{1}'' must be an aggregate expression or appear in GROUP BY clause")
+    String unsupportedNoGroupBySubquery(String a0, String a1);
+
     @BaseMessage("Exist must have exact one subquery")
     String canOnlyOneExistSub();
 

--- a/fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java
+++ b/fe/fe-core/src/test/java/com/starrocks/sql/plan/SubqueryTest.java
@@ -12,7 +12,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-
 package com.starrocks.sql.plan;
 
 import com.starrocks.common.FeConstants;
@@ -1874,5 +1873,18 @@ public class SubqueryTest extends PlanTestBase {
                 " else null end end from tmp a;";
         Pair<String, ExecPlan> pair = UtFrameUtils.getPlanAndFragment(connectContext, sql);
         assertContains(pair.first, "CTEAnchor(cteid=2)");
+    }
+
+    @Test
+    public void testHavingSubqueryNoGroupMode() {
+        String sql = "select v3 from t0 group by v2 having 1 > (select v4 from t1 where t0.v1 = t1.v5)";
+        long sqlMode = connectContext.getSessionVariable().getSqlMode();
+        try {
+            connectContext.getSessionVariable().setSqlMode(0);
+            Assert.assertThrows("must be an aggregate expression or appear in GROUP BY clause", SemanticException.class,
+                    () -> getFragmentPlan(sql));
+        } finally {
+            connectContext.getSessionVariable().setSqlMode(sqlMode);
+        }
     }
 }

